_posts/2019-03-18-speed-up-your-pipeline-using-cache.md

```diff
@@ -0,0 +1,119 @@
+image fast. We started out with a buildspec file that looked like so:
+```yaml
+env:
+  variables:
+     RACK_ENV: "test"
+
+phases:
+  install:
+    commands:
+      - docker run -d --rm -p 5432:5432 -e POSTGRES_USER=my-application_test -e POSTGRES_PASSWORD=postgres -e POSTGRES_DB=test-db postgres:alpine
+      - curl -sL https://deb.nodesource.com/setup_10.x | bash -
+      - apt-get install -y nodejs
+
+  pre_build:
+    commands:
+      - bundle install
+
+  build:
+    commands:
+      - rake db:test:prepare
+      - rspec spec
+```
+As can be seen in the file, we simply ran an AWS codebuild image based on the ruby image, spun up a
+postgres database, installed node, installed the gems and ran the tests.
+
+By taking advantage of packaging the application in a docker container using a separate dockerfile,
+we ended up with a config that was as follows:
+```yaml
+version: 0.2
+
+env:
+  variables:
+     REPO_URL: <Account ID>.dkr.ecr.ap-southeast-2.amazonaws.com/my-application
+     REF_NAME: my-application-tests
+
+phases:
+  install:
+    commands:
+      - $(aws ecr get-login --no-include-email --region ap-southeast-2)
+      - docker pull $REPO_URL:$REF_NAME || true
+
+  pre_build:
+    commands:
+      - docker build --file Dockerfile-testing --cache-from $REPO_URL:$REF_NAME --tag $REPO_URL:$REF_NAME --tag my-application-tests .
+
+  build:
+    commands:
+      - docker-compose --file docker-compose-testing.yml run rspec
+
+  post_build:
+    commands:
+      - docker push $REPO_URL
+```
+And a docker-compose file:
+```yaml
+version: '3'
+services:
+  postgres:
+    image: postgres:alpine
+    ports:
+      - 5432
+    environment:
+      POSTGRES_USER: my-application_test
+      POSTGRES_PASSWORD: postgres
+      POSTGRES_DB: test-db
+
+  rspec:
+    image: my-application-tests
+    volumes:
+      - .:/my-application
+    command: "./run-tests.sh"
+    depends_on:
+      - postgres
+    environment:
+      DB_HOST_URL: postgres
+```
+For rails, we use RSpec as the test runner. The script specfied in the docker-compose file is:
+```bash
+#!/bin/sh
+bundle exec rake db:test:prepare && bundle exec rspec spec
+```
+
+## The Result
+
+At a technical level, docker builds up containers using layers. Those layers are hashed, and then
+the image is stored on AWS ECR. When a new commit is made, the source is pulled in and we specify an
+image to use as our cache to build from. When the container goes to build it can reuse many of the
+layers, as the image of the container only changes when new dependencies (read: gems) are added to
+the image. The result is a fast feedback loop for the developer. After completing their work, and
+validating that it does just work on their machine, they're able to get a quick response from the CI
+tooling without having to wait around with enough time for the focus to fade or distractions to come
+into play.
+
+The proof is in the pudding:
```

Comment:
What next? What should the reader do with this information?

Reply:
Added something, thanks